### PR TITLE
Add an HDC api endpoint to handle chart requests

### PR DESCRIPTION
**Pull request by reed1**
This fixes issue #791 

How to test the feature:

- Set HDC_TOKEN env in fastapi
- Point the chart_data url to "/hdc" of fastapi server instead of api.earthobservation.vam.wfp.org (changes already included in this PR)

Only warning (not error) will show if the HDC_TOKEN is not set. I saw bug #839, so I thought that should apply to backend as well

### CODE REVIEW

**Comment by ericboucher** [comment]
@wadhwamatic would you like us to deploy the backend part of this PR for testing?

**Comment by wadhwamatic** [comment]
Yes please!

File: frontend/src/config/cambodia/layers.json [hunk 1]
```diff
@@ -227,7 +227,7 @@
       "styles": "rfh_16_0_300"
     },
     "chart_data": {
-      "url": "https://api.earthobservation.vam.wfp.org/stats/admin/fetch",
+      "url": "https://prism-api.ovio.org/hdc",
```
Comment (ericboucher) [inline frontend/src/config/cambodia/layers.json:230]: Now that the data source is going to be the same across countries, I think it might be worth extracting into a config variable as "chart_api" or similar and make it a default.

This way you can remove "chart_data.url" from all the layers.json. But keep it available in case we need to access charts from other sources.

File: frontend/src/context/datasetStateSlice.ts [hunk 2]
```diff
@@ -247,7 +247,7 @@ export const loadAdminBoundaryDataset = async (
 
   const hdcRequestParams = {
     level,
-    admin_id: adminCode,
+    id_code: adminCode,
```
Comment (ericboucher) [inline frontend/src/context/datasetStateSlice.ts:250]: Is this change necessary? Do we know what the most common standard is?
admin_id vs id_code?

Linked to my comment on URLs, ideally the frontend stays indifferent as to where the data is coming from.

**Comment by reed1** [comment]
@ericboucher I updated the branch based on your suggestions. I put chartApiUrl below serversUrls in prism.json. Also restored client request with admin_id.

File: frontend/src/config/afghanistan/prism.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
   "serversUrls": {
     "wms": ["https://api.earthobservation.vam.wfp.org/ows/wms"]
   },
+  "chartApiUrl": "https://prism-api.ovio.org/hdc",
```
Comment (ericboucher) [inline frontend/src/config/afghanistan/prism.json:16]: If this is going to be the same for every country, this could even live in a constants.ts, similarly to what's done in the PR linked above

File: api/app/hdc.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+"""Get data from HDC"""
```
Comment (ericboucher) [inline api/app/hdc.py:1]: Add a comment explaining what HDC is

**Comment by ericboucher** [comment]
The API part looks good, I'll deploy now so that we can test.

File: api/app/main.py [hunk 5]
```diff
@@ -176,6 +177,11 @@ def get_acled_incidents(request: Request):
     return Response(content=response.content, media_type="application/json")
 
 
+@app.get("/hdc")
+def wrap_get_hdc_stats(request: Request):
```
Comment (ericboucher) [inline api/app/main.py:181]: you need to better type this request

**Comment by reed1** [comment]
@ericboucher if possible, can you forward me error log for 

https://prism-api.ovio.org/hdc?level=0&admin_id=170&coverage=full&vam=rfh&start=2022-06-07&end=2023-06-07

If it's requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://api.wfp.org/hdc/1.0.0/stats/admin?level=0&id_code=170&coverage=full&vam=rfh&start=2022-06-07&end=2023-06-07
ERROR:uvicorn.error:Exception in ASGI application

Then the HDC_TOKEN env is invalid or not set. I've sent you the token in Slack, let me know if that works, thank you

**Comment by ericboucher** [comment]
Some of the charts are not working it seems. I get the following logs:

```
api        | requests.exceptions.HTTPError: 422 Client Error: UNPROCESSABLE ENTITY for url: https://api.wfp.org/hdc/1.0.0/stats/admin?level=0&id_code=170&coverage=full&vam=rfb&start=2022-06-07&end=2023-06-07
```

Screenshot 2023-06-07 at 10 11 16 AM

**Comment by reed1** [comment]
Confirmed, the rfb doesn't work,

```
  // HDC API expects a parameter which depends on the layer
  // rainfall = rfh; ndvi = vim; blended = rfb
  const getVamParam = () => {
    switch (true) {
      case serverLayerName.includes('vim'):
      case serverLayerName.includes('viq'):
        return 'vim';
      case serverLayerName.includes('blended'):
        return 'rfb';
      default:
        return 'rfh';
    }
  };
```

vim and rfh work, but rfb doesn't. @wadhwamatic perhaps you can notify the API owner on HDC.

curl -H "Authorization: Bearer TOKEN" 'https://api.wfp.org/hdc/1.0.0/stats/admin?level=0&id_code=170&coverage=full&vam=rfh&start=2022-06-07&end=2023-06-07'

does not works, it gives:
```
{"detail":[{"ctx":{"enum_values":["rfh","vim"]},"loc":["query","vam"],"msg":"value is not a valid enumeration member; permitted: 'rfh', 'vim'","type":"type_error.enum"}]}
```

**Comment by wadhwamatic** [comment]
@reed1 - I sent a message to Valentin copying you. We should hear back soon.

**Comment by wadhwamatic** [comment]
@reed1 - sorry I wasn't following this very closely. I see an issue with how we've implemented this. It seems we're expecting that there can only be 1 source for chart data for a deployment. I think we need to keep the option to configure charts / layers which can get data from additional sources, not just the one set in constants.ts.

Right now, we're seeing the issue that 'rfb' is not yet available through the WFP Gateway but still exists through the previous HDC API. If we can allow for a default config which covers all charts but rfb, but still be able to configure rfb using the old API, that would be ideal. That should also allow us to have more than 1 chart data source in the future.

Let me know what you think, also @ericboucher

**Comment by ericboucher** [comment]
@wadhwamatic this is how it's done yes, each layer can override the source to use a different URL

File: frontend/src/utils/admin-utils.ts [hunk 6]
```diff
@@ -50,7 +51,10 @@ export const getChartAdminBoundaryParams = (
 ): AdminBoundaryParams => {
   const { serverLayerName, chartData } = layer;
 
-  const { levels, url, fields: datasetFields } = chartData!;
+  const { levels, url: chartUrl, fields: datasetFields } = chartData!;
+
+  // Take in chartUrl if provided, otherwise use default CHART_API_URL
+  const url = chartUrl || CHART_API_URL;
```
Comment (ericboucher) [inline frontend/src/utils/admin-utils.ts:57]: cc @wadhwamatic this is where we select the URL from layers.json, or default to CHART_API_URL otherwise

**Comment by ericboucher** [comment]
@reed1 alsmot there :)
Please add typing for the API endpoint and as Amit suggested, revert to the old URL for`rfb` datasets for now.

**Comment by reed1** [comment]
@ericboucher Or maybe route rfb in the backend to 'api.earthobservation.vam.wfp.org' ? That way if the HDC is ready then only backend needs to be updated, no need to rebuild frontend instances. The only custom URL for now is in Ukraine rainfall_dekad: https://dev.earthobservation.vam.wfp.org/prism/ukr.

**Comment by ericboucher** [comment]
@reed1 I'd rather make this change in the frontend but I'll let @wadhwamatic decide if it's too cumbersome.

**Comment by wadhwamatic** [comment]
> @reed1 I'd rather make this change in the frontend but I'll let @wadhwamatic decide if it's too cumbersome.

@ericboucher @reed1 - yes, I'd also prefer we handle this as changes to the frontend

**Comment by reed1** [comment]
Sure, configs changed for blended charts:

rfb_blended_moz_dekad
rfq_blended_moz_dekad
rfb_blended_zwe_dekad
rfq_blended_zwe_dekad

File: api/app/main.py (hunk 5 above)
Comment (ericboucher) [reply api/app/main.py:181]: @reed1 can you take care of this so that we can merge? Thanks.

Comment (ericboucher) [reply api/app/main.py:181]: You need to type the request, like the one for kobo/dates. This will allow the API to parse the request and let the user know what parameters are missing, if any. While you're at it, maybe you can type the ACLED request as well? Let me know if you run into any issues

```
@app.get("/kobo/dates")
def get_kobo_form_dates(
    koboUrl: HttpUrl,
    formId: str,
    datetimeField: str,
    filters: Optional[str] = None,
):
```

Comment (reed1) [reply api/app/main.py]: @ericboucher , updated, let me know if that works for you. I tried to use Depends to avoid duplicating function signature. It works on POST, but not GET, only on recent version of fastapi it will work on GET.

**Review by ericboucher — APPROVED** [review]

**Comment by ericboucher** [comment]
I am redeploying the latest version of the API so we can test

File: api/app/main.py [hunk 7]
```diff
@@ -161,19 +163,49 @@ def stats(stats_model: StatsModel) -> list[dict[str, Any]]:
...
     response = get(acled_url, params=params.dict())
     response.raise_for_status()
 
-    return Response(content=response.content, media_type="application/json")
+    return JSONResponse(content=response.content)
+
+
+@app.get("/hdc")
+def wrap_get_hdc_stats(
+    level: str, admin_id: str, coverage: str, vam: str, start: str, end: str
```
Comment (ericboucher) [inline api/app/main.py:197]: non of the params are optional?

File: api/app/main.py [hunk 8]
```diff
@@ -161,19 +163,49 @@ def stats(stats_model: StatsModel) -> list[dict[str, Any]]:
...
+    iso: int,
+    limit: int,
+    fields: Optional[str],
+    email: str,
+    key: str,
+    event_date: Optional[date],
+):
     acled_url = "https://api.acleddata.com/acled/read"
 
     try:
```
Comment (ericboucher) [inline api/app/main.py:176]: I don't think you need this try/except anymore since the typing you introduced should take care of this now.

**Review by wadhwamatic — APPROVED** [review]
looks good from my side too - thanks @reed1 !
